src/threads_mngr.c

```diff
@@ -130,54 +119,40 @@ __attribute__ ((noinline))
 static void invoke_callback(int sig) {
     UNUSED(sig);
 
-    /* If the lock is already locked for write, we are running cleanups, no reason to proceed. */
-    if(0 != pthread_rwlock_tryrdlock(&globals_rw_lock)) {
-        serverLog(LL_WARNING, "threads_mngr: ThreadsManager_cleanups() is in progress, can't invoke signal handler.");
-        return;
-    }
-
-    if (g_callback) {
-        g_callback();
-        size_t curr_done_count;
-        atomicIncrGet(g_num_threads_done, curr_done_count, 1);
-
-        /* last thread shuts down the light */
-        if (curr_done_count == g_tids_len) {
-            sem_post(&wait_for_threads_sem);
-        }
-    }
-
-    pthread_rwlock_unlock(&globals_rw_lock);
+    g_callback();
+    atomicIncr(g_num_threads_done, 1);
 }
 
 static void wait_threads(void) {
-    struct timespec ts;
-    clock_gettime(CLOCK_REALTIME, &ts);
+    struct timespec timeout_time;
+    clock_gettime(CLOCK_REALTIME, &timeout_time);
 
     /* calculate relative time until timeout */
-    ts.tv_sec += RUN_ON_THREADS_TIMEOUT;
+    timeout_time.tv_sec += RUN_ON_THREADS_TIMEOUT;
+
+    /* Wait until all threads are done to invoke the callback or until we reached the timeout */
+    size_t curr_done_count;
+    struct timespec curr_time;
 
-    int status = 0;
+    do {
+        atomicGet(g_num_threads_done, curr_done_count);
+        clock_gettime(CLOCK_REALTIME, &curr_time);
+    } while (curr_done_count < g_tids_len &&
+            curr_time.tv_sec <= timeout_time.tv_sec);
 
-    /* lock the semaphore until the semaphore value rises above zero or a signal
-    handler interrupts the call. In the later case continue to wait. */
-    while ((status = sem_timedwait(&wait_for_threads_sem, &ts)) == -1 && errno == EINTR) {
-        serverLog(LL_WARNING, "threads_mngr: waiting for threads' output was interrupted by signal. Continue waiting.");
-        continue;
+    if (curr_time.tv_sec > timeout_time.tv_sec) {
+        serverLogFromHandler(LL_WARNING, "wait_threads(): waiting threads timed out");
     }
+
 }
 
 static void ThreadsManager_cleanups(void) {
-    pthread_rwlock_wrlock(&globals_rw_lock);
```

Comment:
right. I increased the timeout to 10 s.

Reply:
i don't think that's a valid solution, first, that long timeout can cause issues when one thread hangs, but secondly, it doesn't really solve the problem.

let's see what's the worse that can happen.
in the past it could have cause memory corruption, and now it'll just mess up the next trace (only in case two are executed in close proximity).
is that right?

maybe we can attempt to reproduce that and see if there aren't other issues?
i.e. reduce the timeout, slow down the threads, and trigger many traces repeatedly.

Reply:
there are some scenarios if some thread is hanging:
1. main thread reached timeout, nullified the callback pointer
one of the threads is trying to invoke `g_callback()`, that's a segfault 
2. main thread reached timeout, we set all the ThreadsManager globals to 0 and returned from ThreadsManager_runOnThreads, we are done printing the stacktrace and returned from `writeStacktraces`
one of the threads is now managed to call the callback, but then hanged for some reason, now it is writing the stacktrace to the an array that doesn't exist anymore (it was on `writeStacktraces` stack). also, it will increase`g_num_threads_done` in the threads mngr, and the next call will start from 1 (this is easy to fix by setting this variable to 0 at the beginning as well) 
3.main thread reached timeout, and we now print a stacktrace while a thread writes to this exact location in the array (not very likely but can happen). we will get some corrupted print of the stack trace, or of the thread's name, depending on where in `collect_stacktrace_data` and `writeStacktraces` were at this point.

All scenarios are very unlikely and I think the only way to avoid them is to lock critical code sections unless you have a better idea.

Reply:
1. let's keep nullifying the callback, and change the code to print some log warning and exit if it's null.
2. setting it to 0 on startup seems good. i suppose it can still fire after we set it to 0, but the only damage is that one one thread will be printed twice, or that another one would be missing. i think that's ok.
3. i suppose such corruption will only result in a messed up stack trace, but nothing fatal.

these are all unlikely, so i don't care much about the outcome, as long as it doesn't cause any crashes or memory corruptions.

Reply:
1. One problem is that checking and invoking is not one atomic operation
2. we have two counters to consider: one is in the threads mngr, which counts how many threads are done with invoking the callback. if a thread returns from the callback after a timeout and increases this counter, nothing will happen in the next call to the thread manager (if we fix setting this counter to 0). but in what scenario do we want to print the stacktrace that often? The second counter is in debug.c and we use it to take a memory pool from the array allocated in `writeStacktraces` before calling the threads_mngr. if we already returned from `writeStacktraces` and the thread trying to get this memory or already holds one, it will access a memory that doesn't exist anymore. But again, why would a thread invoke the callback successfully and then get stuck for a long time and then suddenly continue?  
3. yes

Reply:
regarding 2- writing to a stack memory that doesn't exist anymore will not cause a crash it won't affect the next calls to `writeStacktraces` (unless we call it again and the thread is still working on invoking the callback)

Reply:
1. we can fetch the pointer to the stack (most likely the compiler will store it in the register anyway), and then test it before calling.
2. why is not an interesting question, i don't want a long timeout (specifically when this doesn't come from a fatal crash), ideally it should all be very quick (<10ms), and i certainly don't wanna hold the process for more than a second when some thread misbehaves. now considering a short timeout, even if the threads don't misbehave, there could always be timing issues, and any scenario that's theoretically possible, will happen at some point, and we should at least make sure we don't crash or corrupt heap.

if we can't handle the above cases. another way out could be to just prevent the next stack trace collection, if the previous one didn't complete. this means that if one thread blocks forever, we'll only get one partial trace, and next ones will be skipped.
if a thread is just late, then it'll be excluded from the trace, but won't block the next trace.
WDYT?

Reply:
1. ok
2. Currently it's 10. we saw that How do you differentiate between a stuck thread and a late one?

Reply:
and we can't corrupt the heap cause we don't use any malloc's

Reply:
i didn't suggest to differentiate between a stack thread and a late one, i just said these are two cases we need to handle.
stack corruption can lead to heap corruption and vice versa.. anyway, so you say that the one leading to a segfault is now handled, and the other ones will just lead to bad stack trace in the log, but no effects on the other mechanisms in redis.
if that's the case then i'm fine with it. but if we're afraid of unexpected issues, i'm also ok in preventing any stack trace collection until the previous one ended (if the hangs indefinitely, then the whole thing will get disabled till redis is restarted).

Reply:
Following the discussion with @oranagra and @yossigo 
**Background:** `writeStacktraces` allocates on its stack a buffer of size sizeof(stacktrace_data) * len_tids, where stacktrace_data =~ 800 b and len_tids is the current number of threads willing to receive SIGUSR2 signal.
It saves the pointer in a global variable.
Then it calls the thread_mngr that sends SIGUSR2 to all `len_tids` threads, and each one invokes `collect_stacktrace_data`. In this callback, each threads gets its unique spot in the array allocated by `writeStacktraces` and writes its stacktrace into it.
When all the threads are done, or we reach timeout we return from thread mngr, print the stacktraces we were able to collect and the stack allocation is no longer available.
**The problem**: one or more of the threads are late, we reach timeout, return from the thread mangr, exit `writeStacktraces`, and then the thread wakes up and starts writing to this address and corrupts other function stack.

**One solution** that was suggested was to pre-allocate the stacktrace_data array and save it to a global pointer that will live as long as the server is up. The of such array with our current limits (maximal threads number = 50 and max address per thread = 100) is ~40kb. Today an empty instance of Redis is ~800kb, and 40kb is pretty big.
We can shrink it if we decide to allow a smaller backtrace depth.

**Changing this allocation size dynamically** requires syncing and making sure there are no threads currently writing to it, and we don't need any additional struggles with sync mechanisms...

The best solution that is both static and doesn't increase the memory significantly is **using mmap**, however  we don't know how `backtrace_symbols_fd` behaves when called a synchronically.

**writing directly to the log file** was considered but backtrace_symbols (returns a malloc'd array of translated stacktrace address as strings) is obviously not AS-safe.

Reply:
@tezc please have a look :point_up:

Reply:
I like option-3. 

Currently, we already call `backtrace()` which is listed as not async-safe (if this function call triggers loading libbacktrace.so to memory). We are already taking this risk. We can also call `backtrace_symbols_fd()` in threads. It's listed as "AS-Safe" here: [link](https://www.gnu.org/software/libc/manual/html_node/Backtraces.html)

> we'll also have issues with the other per thread info we wanna print..

@meiravgri What are these issues? Maybe we can find a workaround for it? WDYT?

---

Passing fd to threads to make them print their own stacktraces makes sense to me. Though, I assume this scenario can occur:

1- Main thread opens log fd.
2- One thread is late but main thread moves on, prints stacktraces to the log file and close fd.
3- Later, same fd number is assigned to something else in the process.
4- Slow thread writes to this fd. 

We need some sort of synchronization around fd. Maybe you can think of something better but first thing that comes to my mind, we can create a pipe() on process startup and let threads write to it. Then, main thread copies pipe content to the actual log file:

```c
int pipe_fds[2];
int init_on_process_startup()
{
       // Non-blocking pipe
       anetPipe(pipe_fds, O_CLOEXEC|O_NONBLOCK, O_CLOEXEC|O_NONBLOCK);
}

void writeStacktraces(int log_fd);
{
      // A "slow thread" can write something after main thread timeouts in the prev round. Just read the pipe to clear it.
      clearPipe(); 
      
      runOnThreads(...); // threads will always write stacktraces to the pipe_fds[1] by calling backtrace_symbols_fd();
      waitForThreads(....);
      readFromPipeAndWriteToLogFd();
}

```

Reply:
> > we'll also have issues with the other per thread info we wanna print..
> 
> @meiravgri What are these issues? Maybe we can find a workaround for it? WDYT?

the issue is that the other info we print could easily get out of order when many threads do several writes each to the same log file.

Reply:
Oh, right... :fearful: 

Maybe threads can fill `struct stacktrace_data` and write it to the pipe fd. Main thread will read structs from the pipe and print it to the log file. Read/write smaller than `PIPE_BUF` is atomic. So, I assume we won't see partial struct in the pipe.

Reply:
write() fails when trying to write to a closed fd, so no need to sync.

Reply:
I mean same fd number can be assigned to some other file or socket after closing the log file. Then, write() will not fail. 

Say fd=15 was log file, we close it and move on. Later, main thread opens another file/socket for some other purpose. OS can assign 15 to this fd (because 15 is available after close()). Then, late thread wakes up and write() to fd=15 successfully. (it will write to a socket or a file mistakenly). I assume this is a possible scenario.

Reply:
> Maybe threads can fill `struct stacktrace_data` and write it to the pipe fd

ok, so we'll only call `backtrace` (not `backtrace_symbols*`) from the thread, but instead of filling a global data structure, we'll pass the data to the caller via pipe, and rely on `write` being atomic?
sounds like a good idea.
then the caller can keep using a stack allocation, not a heap one.
in order to avoid that problem with a closed fd, we can keep that pipe forever instead of re-creating it each time.
WDYT?

Reply:
let me make sure i understood correctly:
with server startup:
create a global pipe

```
__attribute__ ((noinline)) static void collect_stacktrace_data(void) {
    /* allocate stacktrace_data struct */
    stacktrace_data trace_data = {0};

    /* Get the stack trace first! */
    trace_data.trace_size = backtrace(trace_data.trace, BACKTRACE_MAX_SIZE);

    /* get the thread name */
    prctl(PR_GET_NAME, trace_data.thread_name);

    /* get the thread id */
    trace_data.tid = syscall(SYS_gettid);

   writeToPipe(trace_data);


}


void writeStacktraces(int log_fd) {
      // A "slow thread" can write something after main thread timeouts in the prev round. Just read the pipe to clear it.
      clearPipe(); 
      
      runOnThreads(...); // threads will always write stacktraces to the pipe_fds[1] by calling backtrace();
      // we don't need to wait, runOnThreads does that
      readFromPipeAndWriteToLogFd(); // calls to backtrace_symbols_fd()
}
```

Reply:
@oranagra, @meiravgri yes, exactly like that. 

Then, caller can even read one by one: 

```c
void readFromPipe() 
{
    while (1) {
       struct stacktrace stacktrace;
       
       int r = read(pipe_fds[0], &stacktrace, sizeof(stacktrace));
       if (r == sizeof(stacktrace)) {
            write(fd, "thread name, thread id ..."); // log metadata about thread
            backtrace_symbols_fd(stacktrace.trace, ..., fd);
            ...
       }
    }
}

```

This way we avoid any memory synchronization issues (hope we don't miss anything else). Downside is we have to keep a pipe open forever.

Reply:
Just out of curiosity, What are the consequences other than 2 occupied fds? Does it have any impact on the memory consumption ?

Reply:
The overheads are negligible. We have other piles, both for the entire duration of the execution, and temporary ones for a certain purpose.

Reply:
pipe has a buffer for unread data internally, typically 64 kb. OS can do some optimizations though: allocate it on the first use or free it when all data read etc.

Reply:
@tezc 
```
anetPipe(pipe_fds, O_CLOEXEC|O_NONBLOCK, O_CLOEXEC|O_NONBLOCK);
```
regarding the flags you suggested
**`O_CLOEXEC`** - what does it do? if i understood correctly, the pipe will be closes automatically **for the child process** upon `execve()` ?
I agree we need **`O_NONBLOCK`** - don't block (fail) if we:
1. read and the pipe is empty
2. we try to write and there is no space available

I suggest we use also
**`O_DIRECT`** (since Linux 3.4): Each write to the pipe is dealt with as a separate packet, and reads from the pipe will read one packet at a time.

Reply:
`O_CLOEXEC`: Yes, it will automatically close `fds` after exec.. calls. AFAIK, it is a good practice to open flags with this flag by default, otherwise, after an execve() call, fds will be a leak for the new process (because signal handler won't be usable after an execve() call anyway). 

`O_NONBLOCK`: Yes, we can use it just to prevent blocking threads

`O_DIRECT`: As far as I understand, `O_DIRECT` is useful if you are writing variable length messages to the pipe. As we write and read same struct from the pipe, it shouldn't change anything for us. pipe() guarantees atomicity, so we should never see partial struct bytes in the pipe. write() calls should write struct fully or fail without writing a single byte if remaining space in pipe buffer is smaller than sizeof(stacktrace).

If this is not going to change anything for us, I think we should not use it so, our code can stay compatible with older linux distributions. Do you see any other issue? Maybe I miss something.

Reply:
no ,i thought O_DIRECT guerentees atomicity, thanks for the clarification.

The pipe was a great idea IMO. well done
I've implemented it, please take a look @oranagra @tezc @yossigo 

daily tests:
https://github.com/meiravgri/redis/actions/runs/6919472058